extensions/chathistory.md

```diff
@@ -0,0 +1,115 @@
+
+If a nickname is given as the `target` then the server SHOULD include history sent between both the current user and the target nickname as to give a full conversation. The server SHOULD attempt to include history involving other nicknames if either the current user or the target nickname has changed during the requested timeframe.
+
+#### Subcommands
+
+The following subcommands are used to describe how the server should return messages relative to the `timestamp` or `msgid` given.
+
+#### `BEFORE`
+
+    CHATHISTORY BEFORE <target> <timestamp=YYYY-MM-DDThh:mm:ss.sssZ | msgid=1234> <limit>
+    
+Request up to `limit` number of messages before and including the given `timestamp` or `msgid`. Only one timestamp or msgid MUST be given, not both.
+
+#### `AFTER`
+    CHATHISTORY AFTER <target> <timestamp=YYYY-MM-DDThh:mm:ss.sssZ | msgid=1234> <limit>
+Request up to `limit` number of messages after and excluding the given `timestamp` or `msgid`. Only one timestamp or msgid MUST be given, not both.
+
+#### `LATEST`
+    CHATHISTORY LATEST <target> <* | timestamp=YYYY-MM-DDThh:mm:ss.sssZ | msgid=1234> <limit>
+Request the most recent messages that have been sent after and excluding the given `timestamp` or `msgid`. If a `*` is given instead of a timestamp or msgid, the server MUST use the current time as a timestamp. The number of messages returned MUST be equal to or less than `limit`. If a `*` is not given, only one timestamp or msgid MUST be given, not both.
+
+This is useful for retrieving the latest conversation when first joining a channel or opening a query buffer.
+
+#### `AROUND`
+    CHATHISTORY AROUND <target> <timestamp=YYYY-MM-DDThh:mm:ss.sssZ | msgid=1234> <limit>
+Request a number of messages before and after the `timestamp` or `msgid` with the total number of returned messages not exceeding `limit`. The implementation may decide how many messages to include before and after the selected message. Only one timestamp or msgid MUST be given, not both.
+
+This is useful for retrieving conversation context around a single message.
+
+#### `BETWEEN`
+    CHATHISTORY BETWEEN <target> <timestamp=YYYY-MM-DDThh:mm:ss.sssZ | msgid=1234> <timestamp=YYYY-MM-DDThh:mm:ss.sssZ | msgid=1234> <limit>
+Request up to `limit` number of messages between the given `timestamp` or `msgid` values. The returned messages MUST start from the inclusive first message selector, while excluding and finishing on the second - this may be forwards or backwards in time.
+
+#### Returned message notes
+The returned messages MUST be in ascending time order and the `server-time` tag SHOULD be the time at which the message was received by the IRC server. The `msgid` tag that identifies each individual message in a response MUST be the `msgid` tag as originally sent by the IRC server.
```

Comment:
> the `server-time` tag SHOULD be the time at which the message was received by the IRC server

When it comes to linked servers; does this state that the time will be the one from which I am connected to, or the server that processed the message?

If the former, then a message's "server-time" could change depending on which server I've connected to?

Reply:
I interpreted this as "the server that initially processed the message", i.e., the time tag should be the same across all relays and replays of the message.

Intriguingly, the ambiguity you're identifying exists even outside the context of history, and AFAICT is not clarified by the [latest server-time spec](https://ircv3.net/specs/extensions/server-time-3.2).

Reply:
I was thinking “the server that initially processed the message” too, although that likely means you might not see messages in time order (ascending) when they come from other linked servers live which makes some of the below commands tricky to use.

For example I may receive the following messages, the various nick are connected to different linked servers. doe@aus is notable as they are connected to an Australian IRCd which has the higest lattency:

```
@time=2020-01-23T08:39:42.102Z :bot@network.com PRIVMSG #example :Radio is online
@time=2020-01-23T08:39:42.102Z :bot@network.com PRIVMSG #example :Now playing XYZ
@time=2020-01-23T08:39:42.042Z :dennis@eu PRIVMSG #example :Hello World
@time=2020-01-23T08:39:42.021Z :kyle@us PRIVMSG #example :Hello World
@time=2020-01-23T08:39:41.944Z :doe@aus PRIVMSG #example :Hello World
```

If I disconnected after the first message was received or processed, then reconnect and ask for `CHATHISTORY LATEST` giving the last message timestamp that I’ve seen as `2020-01-23T08:39:42.102Z`. Then I SHOULD receive no additional messages despite never having seen the additional 4 messages. `msgid` lookup has similar problems because the server I connect to may not know the order that I’ve received the messages.

Reply:
Over on https://github.com/prawnsalad/ircv3-specifications/pull/1 we're planning to make most of the commands use timestamps only as parameters. That will allow clients to compensate for skew/latency across the network by adding or subtracting a fuzz duration. (10 seconds seems like a sane value: [Unreal disables linking entirely when it detects 30 seconds of skew](https://www.unrealircd.org/docs/FAQ#Why_correct_time_is_important_and_how_to_fix_your_clock).)

Reply:
In what sense does removing the ability to specify msgids allow clients to do anything?

The situation where it makes sense to use times and a fudge factor is when you're starting to update your existing partial history which came from another server. The rest of the time it's pointless waste.

Either way, with the time order requirement, a distributed implementation that's not strictly serialising has to say inconsistent things, either by returning history in an order it didn't really happen or by returning messages with timestamps that they didn't really have. I could imagine either causing problems for a client that tries to construct a canonical history, but I also just don't want to be required to do extra work to provide less accurate information.

Reply:
I just want to get clear on the terms of this debate. If I understand you correctly, your primary objection is to the requirement that results be returned in time order, *not* to the requirement that queries support timestamps as parameters. As a server vendor, you're going to have to do the work required to answer queries of the form "what messages were sent between 2020-01-27T17:22:09.270Z and 2020-01-27T17:40:09.282Z?" regardless.

Here's my position on this issue:

1. Since there is no global total order on messages, it is unnecessarily difficult for client developers to correctly use msgids in ordering queries. With timestamps, the client and all servers in the network can agree on the intended semantics of the query. (If clients need to support timestamp fudging regardless to handle the case where they reconnect to a different server, in what sense is "waste" occurring? What is being wasted?)
1. The preferred compromise is "returning history in an order it didn't really happen", not sending fake timestamps. The only thing that can cause the timestamp order to contradict the "causal order" of messages is clock skew among servers (not latency). Clock skew among well-managed servers should be on the order of a few milliseconds, so this is a rare edge case.
1. The server-time spec already says "a client SHOULD treat the message as having occurred at the given time instead of its current time": it is implicit in this that the timestamp order overrides the order of message delivery (and I think this is the behavior implemented in many clients that support server-time).